rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -236,11 +301,290 @@ impl Body for PathBody {
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to read line 0 only
+            .length(Length::Exact(line_0.len() as u64))
+            .build()
+            .await?;
+
+        let data = body.collect().await?.into_bytes();
+        let data_str = String::from_utf8(data.to_vec())?;
+
+        assert_eq!(&data_str, line_0);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_supports_offset() -> Result<(), Box<dyn std::error::Error>> {
+        let mut file = NamedTempFile::new()?;
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+
+        write!(file, "{}", line_0)?;
+        write!(file, "{}", line_1)?;
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to skip the first line by using offset
+            .offset(line_0.len() as u64)
+            .build()
+            .await?;
+
+        let data = body.collect().await?.into_bytes();
+        let data_str = String::from_utf8(data.to_vec())?;
+
+        assert_eq!(&data_str, line_1);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_offset_and_length_work_together() -> Result<(), Box<dyn std::error::Error>> {
+        let mut file = NamedTempFile::new()?;
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+        let line_2 = "Line 2\n";
+
+        write!(file, "{}", line_0)?;
+        write!(file, "{}", line_1)?;
+        write!(file, "{}", line_2)?;
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to skip line 0 by using offset
+            .offset(line_0.len() as u64)
+            // We want to read only line 1 and stop before we get to line 2
+            .length(Length::Exact(line_1.len() as u64))
+            .build()
+            .await?;
+
+        let data = body.collect().await?.into_bytes();
+        let data_str = String::from_utf8(data.to_vec())?;
+
+        assert_eq!(&data_str, line_1);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn fsbuilder_with_offset_greater_than_file_length_reads_no_bytes(
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let mut file = NamedTempFile::new()?;
+        let line_0 = "Line 0\n";
+        let line_1 = "Line 1\n";
+
+        write!(file, "{}", line_0)?;
+        write!(file, "{}", line_1)?;
+
+        // Ensure that the file was written to
+        file.flush().expect("flushing is OK");
+
+        let body = FsBuilder::new()
+            .path(&file)
+            // We're going to skip all file contents by setting an offset
+            // much larger than the file size
+            .offset(9000)
```

Comment:
So if the length isn't set at all, we default to `Length::UpTo(actual_file_size)` behavior. Is this consistent?

Reply:
In the case of not setting a length, the past and present behavior are the same. The entire file will be read.